Add a GitHub Action for installing slsa-verifier.

kpk47 opened the pull request:
This is a rewrite of https://github.com/slsa-framework/slsa-verifier/pull/233 in Typescript.

@asraa @ianlewis @laurentsimon

laurentsimon reviewed (commented):
looks good. Left a few comment. I think this is the "allow installation of any version". Heard back from Ian whether he's OK with using the `@v1.2.3` opption instead?

kpk47 commented on actions/installer/src/index.ts:
```diff
@@ -0,0 +1,142 @@
+import * as fs from "fs";
+import * as os from "os";
+
+const LATEST_VERSION = "v1.3.0";
```
That part of the GH context is not available to Javascript actions. This is the subset that's available, and it does not include the action path or action ref: https://github.com/actions/toolkit/blob/main/packages/github/src/context.ts

laurentsimon replied:
You can access everything. 3 Options:
- Either you read the `GITHUB_EVENT_PATH` and parse it yourself. I never remember if this is just the payload or the context
- If the context really is not available, create a default action input and populate it by default to the GH context. See 
https://github.com/ossf/scorecard-action/blob/main/action.yaml#L45-L48, you could populate it with something like https://github.com/slsa-framework/slsa-github-generator/blob/main/.github/workflows/builder_go_slsa3.yml#L257
- Simplest option for our use case, is to read the env variable `GITHUB_ACTION_PATH` https://docs.github.com/en/actions/learn-github-actions/environment-variables#default-environment-variables

laurentsimon replied:
re: breaking workflow. You can list releases and install the one that corresponds to the commit. The main question, I think, is whether we want to allow installing any version, or simply install the on the user pins the Action to.

kpk47 commented:
@laurentsimon I spoke with Ian about pinning by release tags, and he's fine with it. Unfortunately, I've written a bunch of versions of this action and decided that I like the version with a `version` input and an explicit `LATEST_VERSION` in the javascript. Here's my reasoning. LMK if you disagree.

Pros
- Users can still pin to a release tag if they want. The `version` input is optional, and pinning to a release tag will install whichever version was latest at the time of release (i.e. `@1.2.3` will install `v1.2.3`).
- If we have an explicit `LATEST_VERSION` in the code, users can pin to a commit hash. GH's documentation on using public actions recommends pinning to a commit SHA rather than a release version or `main`, so I think it's important we support that use case.
- It's easier to write the action if it takes the version as input. `action_path` and `action_ref` are not part of the subset of the GH context that's available to Javascript actions, so getting the version that way requires an ugly hack with a second (composite) action. I wrote that version and discarded it because my intuition is that it would be brittle.
- It's easier to test the action if it takes the version as an input since we can use a matrix to cover all supported release versions. We cannot parse the version from a local action (it's always `main` at whatever commit you have checked out), so we'd need to use a second repo for running tests. I don't think it's worth the hassle.
- After looking through the GHA Marketplace, I realized that it would be very unusual if we only allowed choosing the installed version by pinning the action to a release tag. Every other `*-setup` action I found either supported taking the version as an input or only supported installing the latest release. I think that being unusual in this way would be helpful for adoption.

Cons
- We have to keep the `LATEST_VERSION` constant up to date, which most likely means adding an extra step to our release process.

laurentsimon commented:
> @laurentsimon I spoke with Ian about pinning by release tags, and he's fine with it. Unfortunately, I've written a bunch of versions of this action and decided that I like the version with a `version` input and an explicit `LATEST_VERSION` in the javascript. Here's my reasoning. LMK if you disagree.
> 
> Pros
> 
> * Users can still pin to a release tag if they want. The `version` input is optional, and pinning to a release tag will install whichever version was latest at the time of release (i.e. `@1.2.3` will install `v1.2.3`).
> * If we have an explicit `LATEST_VERSION` in the code, users can pin to a commit hash. GH's documentation on using public actions recommends pinning to a commit SHA rather than a release version or `main`, so I think it's important we support that use case.

We can support this too.

> * It's easier to write the action if it takes the version as input. `action_path` and `action_ref` are not part of the subset of the GH context that's available to Javascript actions, so getting the version that way requires an ugly hack with a second (composite) action. I wrote that version and discarded it because my intuition is that it would be brittle.

`action_path` is accessible, and it contains the `ref`

> * It's easier to test the action if it takes the version as an input since we can use a matrix to cover all supported release versions. We cannot parse the version from a local action (it's always `main` at whatever commit you have checked out), so we'd need to use a second repo for running tests. I don't think it's worth the hassle.

We can use an API to list the version, and dynamically generate the matrix, as in https://michaelheap.com/dynamic-matrix-generation-github-actions/.

> * After looking through the GHA Marketplace, I realized that it would be very unusual if we only allowed choosing the installed version by pinning the action to a release tag. Every other `*-setup` action I found either supported taking the version as an input or only supported installing the latest release. I think that being unusual in this way would be helpful for adoption.

That's the main argument against.

> 
> Cons
> 
> * We have to keep the `LATEST_VERSION` constant up to date, which most likely means adding an extra step to our release process.

Let's enjoy the week-end. I know re-writing so many times is frustrating. Let's sync next week

kpk47 commented on .github/workflows/test_install.yml:
```diff
@@ -0,0 +1,55 @@
+    runs-on: ubuntu-latest
+    strategy:
+      matrix:
+        verifier-version: [v0.0.1, v1.0.0, v1.0.1, v1.0.2, v1.0.3, v1.1.0, v1.1.1, v1.2.0, v1.3.0, latest]
```
Done

kpk47 commented on actions/installer/dst/index.js:
```diff
@@ -0,0 +1,129 @@
+"use strict";
```
Done

laurentsimon commented on .github/workflows/test_install.yml:
```diff
@@ -0,0 +1,20 @@
+permissions: read-all
```
Can you follow the file naming convention we use on the generator https://github.com/slsa-framework/slsa-github-generator/tree/main/.github/workflows?
Basically have pre-submit.actions.yml for Action presubmit, etc.

We only had a pre-submit.yml in this repo, but looks like it's time to change.

laurentsimon reviewed (approved):
Thanks!
Just added a nit to rename files for pre-submit, and we're good to merge

ianlewis commented on actions/installer/README.md:
````diff
@@ -0,0 +1,37 @@
+    verifier-version: v1.3.0
+```
+
+NB: Pinning this action to a relase tag or commit SHA will pin you to the latest version of `slsa-verifier` at the time of release.
````
```suggestion
NB: Pinning this action to a release tag or commit SHA will pin you to the latest version of `slsa-verifier` at the time of release.
```

BTW: what does NB mean? consider expanding it if it's an acronym.

kpk47 replied:
It's Latin, so expanding the acronym would be kind of weird (https://en.wikipedia.org/wiki/Nota_bene). I've just replaced it with "Note: ..."

ianlewis commented on actions/installer/README.md:
```diff
@@ -0,0 +1,37 @@
+
+### Optional Inputs
+
+| Input | Description |
+| --- | --- |
+| `verifier-release` | `slsa-verifier` version to use. Default value is `latest`. |
```
nit: run the markdown file through prettier before committing.

ianlewis commented on actions/installer/README.md:
```diff
@@ -0,0 +1,37 @@
+
+See https://github.com/slsa-framework/slsa-verifier/releases for the list of available `slsa-verifier` releases.
+
+For a full example workflow, see `.github/workflows/test_installer.yml`.
```
nit: consider making this a link? In order to be a relative URL I guess it would need to be with `../../.github/workflows/test_installer.yml`

ianlewis commented on actions/installer/README.md:
```diff
@@ -0,0 +1,37 @@
+For a full example workflow, see `.github/workflows/test_installer.yml`.
+
+This action requires using GitHub-provided Linux runners.
+
```
nit: extra line at the end.

ianlewis commented on actions/installer/spec/index.test.ts:
```diff
@@ -0,0 +1,56 @@
+const exec = require("@actions/exec");
```
nit: we haven't been good about this but add the copyright header with 'SLSA Authors' as the author.

See the header here for an example: https://github.com/slsa-framework/slsa-github-generator/blob/main/slsa/buildtype.go

same for other code files.

ianlewis commented on actions/installer/src/index.ts:
```diff
@@ -0,0 +1,141 @@
+  try {
+    // Download bootstrap version and validate SHA256 checksum
+    bootstrapVerifierPath = await tc.downloadTool(
+      "https://github.com/slsa-framework/slsa-verifier/releases/download/v1.3.0/slsa-verifier-linux-amd64",
```
nit: use LATEST_VERSION in the url here. Or create another global that sets the bootstrap URL or bootstrap version.

ianlewis commented on actions/installer/src/index.ts:
```diff
@@ -0,0 +1,141 @@
+
+const LATEST_VERSION = "v1.3.0";
+
+const VERIFIER_V130_SHA256 =
```
nit: make the name indicate it's used for bootstrapping. Set another global here to specify the bootstrap version or URL (see other comment).
```suggestion
const BOOTSTRAP_VERIFIER_SHA256 =
```

ianlewis commented on actions/installer/src/index.ts:
```diff
@@ -0,0 +1,141 @@
+  } catch (error: unknown) {
+    const errMsg = error instanceof Error ? error.message : String(error);
+    core.setFailed(`Error executing slsa-verifier: ${errMsg}`);
+    return;
```
nit: is it maybe prudent to call `cleanup` when errors occur as well?

ianlewis reviewed (approved):
Looks good!

laurentsimon commented on actions/installer/README.md:
````diff
@@ -0,0 +1,37 @@
+
+To install a specific version, use:
+```yaml
+uses: slsa-framework/slsa-verifier-installer
````
need to update the doc, since the verifier-version is no longer supported?

laurentsimon reviewed (commented):
I thought we'd use the `GITHUB_ACTION_PATH` to determine the version to download?

kpk47 commented:
@laurentsimon ```GITHUB_ACTION_PATH``` is only available to composite actions (https://docs.github.com/en/actions/learn-github-actions/environment-variables#default-environment-variables).

laurentsimon commented:
> @laurentsimon `GITHUB_ACTION_PATH` is only available to composite actions (https://docs.github.com/en/actions/learn-github-actions/environment-variables#default-environment-variables).

ouch, all these strange corner cases. Good find!
One hack is to declare a composite action and have a single step that calls `node code.js`. Does this work?

Note: GH are working on better APIs for Actions / re-usable workflows to determine their identity / ref, so later we should be able to use it.

kpk47 commented on actions/installer/action.yml:
```diff
@@ -18,5 +18,28 @@ branding:
+        working-directory: actions/installer
+        run: npm ci
+
+      - name: Run build and tests
```
Done

kpk47 commented:
Okay, I think I've address all of your comments. PTAL and then I think we're ready to merge.

laurentsimon commented:
Please re-base

kpk47 commented on actions/installer/README.md:
```diff
@@ -0,0 +1,35 @@
+## Usage
+
+To install the latest version, use:
+
```
Done

kpk47 commented on actions/installer/src/index.ts:
```diff
@@ -0,0 +1,202 @@
+  }
+
+  // If actionRef is a commit SHA, then find the associated version number.
+  const shaRe = /^[a-f\d]{40}$/;
```
Thanks! I copied it from you (https://github.com/slsa-framework/slsa-github-generator/pull/86). 😅

kpk47 commented on actions/installer/action.yml:
```diff
@@ -0,0 +1,48 @@
+      shell: bash
+      run: npm ci
+
+    - name: Run build and tests
```
Done.

kpk47 commented on .github/workflows/pre-submit.actions.yml:
```diff
@@ -0,0 +1,43 @@
+name: Actions pre submits
+on: [pull_request, workflow_dispatch]
+
+permissions: read-all
```
Ack. I'll stash that commit before I rebase so we don't have to rewrite it.

laurentsimon commented on actions/installer/README.md:
````diff
@@ -0,0 +1,29 @@
+uses: slsa-framework/slsa-verifier-installer@v1.3.0
+```
+
+Note: Pinning this action to a release tag or commit SHA will pin you to the latest version of `slsa-verifier` at the time of release.
````
this sentence is no longer true.

kpk47 replied:
Done

laurentsimon commented on actions/installer/README.md:
```diff
@@ -0,0 +1,29 @@
+
+Note: Pinning this action to a release tag or commit SHA will pin you to the latest version of `slsa-verifier` at the time of release.
+
+### Optional Inputs
```
there's no option, so lets remove this as well.

kpk47 replied:
Done

ianlewis commented:
LGTM, pending doc changes that @laurentsimon commented on, and merging the conflicting `pre-submit.yml` changes.

kpk47 commented:
I think I've fixed the merge conflicts, but I can't get GitHub to recognize it. I dropped the changes in `pre-submit.yml` in favor of a straight rename, so I'm not sure what else I can do.

laurentsimon commented:
yeah things get weird when re-basing and files that are deleted have some updates. Have you tried `git rm` the `pre-submit.yml` to see if it helps?

kpk47 commented:
Okay, I think I got it. For posterity, the fix was re-adding `pre-submit.yml`, merging in changes from `main`, removing `pre-submit.cli.yml`, and then re-doing the name change from `pre-submit.yml` to `pre-submit.cli.yml`.

laurentsimon commented on actions/installer/README.md:
```diff
@@ -0,0 +1,21 @@
+
+See https://github.com/slsa-framework/slsa-verifier/releases for the list of available `slsa-verifier` releases.
+
+For a full example workflow, see [../../.github/workflows/test_installer.yml](https://github.com/slsa-framework/slsa-verifier/.github/workflows/test_installer.yml).
```
I think think this file exists anymore, does it?
